PJ05/src/MessageClient.java

```diff
@@ -1,18 +1,417 @@
+                frame.dispose();
+            }
+            if (e.getSource() == okButton) {
+                //TODO send authentication request
+                username = usernameTf.getText();
+                password = passwordTf.getText();
+                authenticateRequest(username, password);
+                if (authenticateResponse()) {
+                    SwingUtilities.invokeLater(new MainInterface());
+                    frame.dispose();
+                } else {
+                    JOptionPane.showMessageDialog(frame, "response message", "Error",
+                            JOptionPane.ERROR_MESSAGE);
+                }
+
+            }
+        };
+        registerButton.addActionListener(actionListener);
+        okButton.addActionListener(actionListener);
 
     }
+
+    public static void main(String[] args) {
+        String hostname;
+        String portString;
+        int port;
+        Socket socket;
+
+
+        try {
+            hostname = JOptionPane.showInputDialog(null, "Please enter the hostname:",
+                    "Connecting...", JOptionPane.INFORMATION_MESSAGE);
+
+            portString = JOptionPane.showInputDialog(null, "Please enter the port number:",
+                    "Connecting...", JOptionPane.INFORMATION_MESSAGE);
+
+            port = Integer.parseInt(portString);
+
+            socket = new Socket(hostname, port);
+
+            if (socket.isConnected()) {
+                SwingUtilities.invokeLater(new MessageClient());
+                bfr = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                pw = new PrintWriter(socket.getOutputStream());
+            }
+
+
+        } catch (NumberFormatException e) {
+            JOptionPane.showMessageDialog(null, "Invalid input!", "Connecting...",
+                    JOptionPane.ERROR_MESSAGE);
+            //return;
+        } catch (IOException e) {
+            JOptionPane.showMessageDialog(null, "Connection failed!", "Connecting...",
+                    JOptionPane.ERROR_MESSAGE);
+            //return;
+        } //connection established
+
+    }
+
+    /**
+     * The method creates an authenticate request to the server
+     * @param username entered by the user
+     * @param password entered by the user
+     */
+    public void authenticateRequest(String username, String password) {
+        Credential credential = new Credential(username, password);
+        AuthenticateRequest authenticateRequest = new AuthenticateRequest(credential);
+        pw.println(authenticateRequest.toString());
+    }
+
+    /**
+     * The method reads the authenticate response
+     * @return true if passes the authentication
+     */
+    public boolean authenticateResponse() {
+        String response;
+        try {
+            response = bfr.readLine();
+        } catch (IOException e) {
+            return false;
+        }
+        return response.equals("true");
+    }
+
```

Comment:
Here's a sample code for a generic request-sending method.

```suggestion
    /**
     * Sends a request to the server and gets the response.
     * @param request The request to be sent to server.
     * @param socket The socket connection already established.
     * @return A response to the request.
     * @throws IOException There's a problem regarding socket connection and sending objects.
     * @throws RequestParsingException The response received cannot be parsed correctly.
     */
    public static Response send(Request request, Socket socket) throws IOException, RequestParsingException, RequestFailedException
    {
        Response response;
        ObjectInputStream objectInputStream = new ObjectInputStream(socket.getInputStream());
        ObjectOutputStream objectOutputStream = new ObjectOutputStream(socket.getOutputStream());

        objectOutputStream.writeObject(request);
        objectOutputStream.flush();
        objectOutputStream.close();

        try{
            response = (Response) objectInputStream.readObject();
        }
        catch(ClassNotFoundException | ClassCastException e){
            e.printStackTrace();
            throw new RequestParsingException();
        }

        if (response.state){
            return response;
        }else{
            if (response.exception != null){
                throw (RequestFailedException) response.exception;
            }
            else {
                throw new RequestFailedException();
            }
        }
    }
```

Note: `RequestFailedException` and `response.exception` are not implemented yet. You need to wait for @fang296 .

Reply:
In order to utilize this function, consider the following code.

```suggestion
    /**
     * Tries to register a new user.
     * @param credential Entered credential for new user.
     * @param profile Entered profile for new user.
     * @return My own UUID.
     * @throws IOException Problem with requesting.
     * @throws RequestParsingException Problem with requesting.
     */
    public UUID register(Credential credential, Profile profile) throws IOException, RequestParsingException{
        RegisterRequest registerRequest = new RegisterRequest(credential, profile);
        try{
            (RegisterResponse) response = send(registerRequest);
            UUID my_uuid = response.uuid;
            System.out.println("register success!");
            return my_uuid;
        }
        catch (UserExistsException e){
            // do something
        }
        catch (InvalidUsernameException e){
            //do something
        }
        catch (InvalidPasswordException e){
            //do something
        }
        catch (RequestFailedException e){
            //This line should not be triggered because the exception must be one of the documented. 
            //You may throw an exception here, or perform some default action.
        }
        return null;
    }
```